contracts/protocol/tokenization/ampl/IncentivizedAAmplERC20.sol

```diff
@@ -0,0 +1,332 @@
+    _approve(_msgSender(), spender, _allowances[_msgSender()][spender].add(addedValue));
+    return true;
+  }
+
+  /**
+   * @dev Decreases the allowance of spender to spend _msgSender() tokens
+   * @param spender The user allowed to spend on behalf of _msgSender()
+   * @param subtractedValue The amount being subtracted to the allowance
+   * @return `true`
+   **/
+  function decreaseAllowance(address spender, uint256 subtractedValue)
+    public
+    virtual
+    returns (bool)
+  {
+    _approve(
+      _msgSender(),
+      spender,
+      _allowances[_msgSender()][spender].sub(
+        subtractedValue,
+        'AAmplERC20: decreased allowance below zero'
+      )
+    );
+    return true;
+  }
+
+  /**
+    wkt, amount = _transferAmount/_totalSupply * totalSupply()
+
+    internal transfer amount for book-keeping is calculated as:
+    _transferAmount = (amount * _totalSupply) / totalSupply()
+
+    _balances[sender] -= _transferAmount
+    _balances[recipient] += _transferAmount
+  */
+  function _transfer(
+    address sender,
+    address recipient,
+    uint256 amount
+  ) internal virtual {
+    require(sender != address(0), 'AAmplERC20: transfer from the zero address');
+    require(recipient != address(0), 'AAmplERC20: transfer to the zero address');
+
+    _beforeTokenTransfer(sender, recipient, amount);
+
+    uint256 currentTotalSupply = totalSupply();
+    uint256 oldSenderBalance = balanceOf(sender);
+    uint256 oldRecipientBalance = balanceOf(recipient);
+
+    uint256 _transferAmount = amount.mul(_totalSupply).div(currentTotalSupply);
+    _balances[sender] = _balances[sender].sub(_transferAmount, 'AAmplERC20: transfer amount exceeds balance');
+    _balances[recipient] = _balances[recipient].add(_transferAmount);
+
+    if(address(_incentivesController) != address(0)) {
+      _incentivesController.handleAction(sender, currentTotalSupply, oldSenderBalance);
+      if (sender != recipient) {
+        _incentivesController.handleAction(recipient, currentTotalSupply, oldRecipientBalance);
+      }
+    }
+  }
+
+
+  /*
+    Before 'amount' is mint to account wkt,
+
+    _balances[account] / _totalSupply = balanceOf(account) / totalSupply()
+
+    We calculate _mintAmount such that the following holds true
+
+    (_balances[account] + _mintAmount) / (_totalSupply + _mintAmount)
+      = (balanceOf(account) + amount)/ (totalSupply() + amount)
+
+    newTotalSupply = totalSupply() + amount
+    newAccountBalance = balanceOf(account) + amount
+    remainingAccountBalance = totalSupply() - balanceOf(account)
+
+    Thus:
+    => _mintAmount = (_totalSupply * newAccountBalance) - (_balances[account] * newTotalSupply) / remainingAccountBalance
+  */
+  function _mint(address account, uint256 amount) internal virtual {
+    require(account != address(0), 'AAmplERC20: mint to the zero address');
+
+    _beforeTokenTransfer(address(0), account, amount);
+
```

Comment:
`getAMPLScalar()` is called multiple times 

`totalSupply` and each `balanceOf` function calls `getAMPLScalar()` and its gest called once in the `mint` function too